src/engine/effects/engineeffectsdelay.cpp

```diff
@@ -0,0 +1,152 @@
+#include "engine/effects/engineeffectsdelay.h"
+
+#include "util/sample.h"
+
+void EngineEffectsDelay::setDelay(unsigned int delaySamples) {
+    m_delaySamples = delaySamples;
+}
+
+void EngineEffectsDelay::process(const CSAMPLE* pIn, CSAMPLE* pOutput, const int iBufferSize) {
+    if (m_oldDelaySamples == 0 && m_delaySamples == 0) {
+        if (pIn != pOutput) {
+            SampleUtil::copy(pOutput, pIn, iBufferSize);
+        }
+
+        return;
+    }
+
+    if (m_oldDelaySamples == m_delaySamples &&
+            m_delayBufferSize == iBufferSize + m_delaySamples) {
+        int delaySourcePos = (m_delayPos + m_delayBufferSize - m_delaySamples) % m_delayBufferSize;
```

Comment:
I don't understand this case. Is it when the delay hasn't changed and the buffer is large enough? Why do we need this exception?

Reply:
It's alright, the second condition part (m_delayBufferSize == iBufferSize + m_delaySamples) was added, because it looked to me not so clean to rely on that the iBufferSize, which is the input parameter, is always the same. If it wouldn't the same, the m_delayBufferSize would contain a bad value. Anyway, the buffer size is set now in the new version in the constructor and is known in compile time, so the condition for this part will be that just the delay hasn't changed (m_oldDelaySamples == m_delaySamples).